plutus-core/plutus-core/src/PlutusCore/Builtin/Meaning.hs

```diff
@@ -73,12 +73,14 @@ class (Typeable uni, Typeable fun, Bounded fun, Enum fun, Ix fun) => ToBuiltinMe
     -- | The @cost@ part of 'BuiltinMeaning'.
     type CostingPart uni fun
 
+    data BuiltinVersion fun
+
     -- | Get the 'BuiltinMeaning' of a built-in function.
-    toBuiltinMeaning :: HasMeaningIn uni val => fun -> BuiltinMeaning val (CostingPart uni fun)
+    toBuiltinMeaning :: HasMeaningIn uni val => BuiltinVersion fun -> fun -> BuiltinMeaning val (CostingPart uni fun)
```

Comment:
It strikes me that part of the reason this PR is more complex than we maybe expected is because this change actually gives us way more power than we want: it lets us change *all* of the builtin meaning, *including the type signature*. We just want to change the runtime behaviour, not the static behaviour! 

All of the changes to compilation and analysis are because we are having to deal with the (unused) possibility that the `BuiltinVersion` might affect the *type* of the builtin.

So here's the question (cc @effectfully ): can we do this in such a way that *only* the runtime behaviour is configurable? I'm not sure it's that easy because we infer the static information (the type scheme) from the runtime information...

Reply:
Here's a possible approach? The downside of this would be we'd actually be looking at the config type at runtime, which seems bad?

```
data BuiltinMeaning val cost config =
    forall args res. BuiltinMeaning
        (TypeScheme val args res)
        ~(config -> FoldArgs args res)
        (BuiltinRuntimeOptions (Length args) val cost)

class (Typeable uni, Typeable fun, Bounded fun, Enum fun, Ix fun) => ToBuiltinMeaning uni fun where
    -- | The @cost@ part of 'BuiltinMeaning'.
    type CostingPart uni fun

    data BuiltinVersion fun

    -- | Get the 'BuiltinMeaning' of a built-in function.
    toBuiltinMeaning :: HasMeaningIn uni val => fun -> BuiltinMeaning val (CostingPart uni fun) (BuiltinVersion fun)
```

Reply:
> It strikes me that part of the reason this PR is more complex than we maybe expected is because this change actually gives us way more power than we want: it lets us change *all* of the builtin meaning, *including the type signature*. We just want to change the runtime behaviour, not the static behaviour! 
> 
> All of the changes to compilation and analysis are because we are having to deal with the (unused) possibility that the `BuiltinVersion` might affect the *type* of the builtin.
> 
> So here's the question (cc @effectfully ): can we do this in such a way that *only* the runtime behaviour is configurable? I'm not sure it's that easy because we infer the static information (the type scheme) from the runtime information...

Yes, the type signature (number of args as well as type of args) may change that is why it is such an invasive change, e.g. the inliner needs the version to figure out the arg length lof a builtin

Reply:
> Nikos just changed Integer to Word8

Ugh, good point. We *could* in principle do this change without changing the type, but it's the cleanest way :/

I guess the thing I feel uncomfortable about is that we could now e.g. compile against BuiltinV1 and then run against BuiltinV2 _and have the optimization we did be incorrect!_. That won't actually happen, of course, because the only respect in which it could screw up optimization is if we changed the arity and we won't do that... but it makes me uncomfortable that we could.

But I buy your argument that currently arity is deeply tied to type, and we are changing the type so... tough :shrug: 

This is going to need a big note though...

> as far as I can see with this PR we do also match on the config at runtime

I guess I thought that when we created the `BuiltinRuntime` we would hopefully end up with a specialized version of the function since the argument is known statically... but we'd need to check.

Reply:
> However there's `RuntimeScheme` stored in a `BuiltinMeaning`, which does give us the arity information that doesn't change across version, so we could try to pull that one out, but I'm not sure if it's worth it.

I take that back, I was going to attempt to get rid of `RuntimeScheme` altogether.

But we can easily add some another `Scheme`. Like tell me what info you want to preserve when a builtin changes and I'll add that to the machinery and you'll be able to rely on it. Plus we don't need any static guarantees and fancy GADTs or whatever, we can just add a test checking that the info that is supposed to be preserved is indeed preserved for each builtin.

> I guess the thing I feel uncomfortable about is that we could now e.g. compile against BuiltinV1 and then run against BuiltinV2 and have the optimization we did be incorrect!. That won't actually happen, of course, because the only respect in which it could screw up optimization is if we changed the arity and we won't do that... but it makes me uncomfortable that we could.

We're changing the behavior of a function in a backwards-incompatible way. All general guarantees go out of the window and trade-offs need to be considered on a case-by-case basis, unless we settle on the specific subset of guarantees and have the machinery for it.

> I guess I thought that when we created the `BuiltinRuntime` we would hopefully end up with a specialized version of the function since the argument is known statically... but we'd need to check.

The `BuiltinRuntime` is stored in the `BuiltinsRuntime` for that particular version, so we'll need to statically allocate the `BuiltinsRuntime` for each version (like with immediate vs deferred unlifting), which we don't do currently. Maybe we should. That would probably amount to unconditionally inlining `mkDynEvaluationContext`?

But maybe we could get away with caching instead of inlining like we do with costing. I'd need to look at the Core to tell.